src/math/BoundedField.cc

```diff
@@ -83,19 +83,110 @@ struct Divide {
 };
 
 template <typename T, typename F>
-void applyToImage(BoundedField const &field, image::Image<T> &img, F functor, bool overlapOnly) {
+void applyToImage(BoundedField const &field, image::Image<T> &img, F functor, bool overlapOnly, int xStep,
+                  int yStep) {
     geom::Box2I region(field.getBBox());
     if (overlapOnly) {
         region.clip(img.getBBox(image::PARENT));
     } else if (region != img.getBBox(image::PARENT)) {
         throw LSST_EXCEPT(pex::exceptions::RuntimeError,
                           "Image bounding box does not match field bounding box");
     }
-    for (int y = region.getBeginY(), yEnd = region.getEndY(); y < yEnd; ++y) {
-        typename image::Image<T>::x_iterator rowIter =
-                img.x_at(region.getBeginX() - img.getX0(), y - img.getY0());
-        for (int x = region.getBeginX(), xEnd = region.getEndX(); x < xEnd; ++x, ++rowIter) {
-            functor(*rowIter, field.evaluate(x, y));
+
+    if (yStep > 1 || xStep > 1) {
+
+        // Bilinear interpolation with no dynamic memory allocation:
+        // iterate over (y0, y1) intervals, then (x0, x1) intervals to evaluate
+        // the field at the corners of each box, then iterate over pixels in
+        // the box (in y, then x) to fill interpolated values.
+        // This means we evaluate the function multiple times at some points
+        // (when we move to a new interval in y, to be specific).
+        // Could probably be optimized for the cases where either step is 1,
+        // but that seems premature.
+
+        // Local function to do bilinear interpolation using points
+        // [(x0, y0), (x0, y1), (x1, y0), (x1, y1)] with values
+        // [z00, z01, z10, z11], in pixels [x0:xEnd, y0:yEnd]
+        auto interpolate = [&img, &functor](
```

Comment:
I feel like `interpolate` and `processRowBlock` are complicated enough that they should be anonymous functions instead of lambdas. I'd strongly make that case in python (or that they at least be inner functions), but I don't know what the standard is in C++, and our dev guide doesn't even mention lambdas.